core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts

```diff
@@ -321,76 +278,108 @@ export class OptimizedSsrEngine {
     options,
     renderCallback,
     request,
-    isFirstRequestForKey,
   }: {
     filePath: string;
     options: any;
     renderCallback: SsrCallbackFn;
     request: Request;
-    isFirstRequestForKey: boolean;
   }): void {
-    const renderingKey = this.getRenderingKey(request);
-
     if (!this.ssrOptions?.reuseCurrentRendering) {
-      this.log(`Rendering started (${request?.originalUrl})`);
-      this.expressEngine(filePath, options, renderCallback);
+      this.startRender({
+        filePath,
+        options,
+        renderCallback,
+        request,
+      });
       return;
     }
 
+    const renderingKey = this.getRenderingKey(request);
     if (!this.renderCallbacks.has(renderingKey)) {
       this.renderCallbacks.set(renderingKey, []);
     }
     this.renderCallbacks.get(renderingKey)?.push(renderCallback);
 
-    if (isFirstRequestForKey) {
-      this.log(`Rendering started (${request?.originalUrl})`);
-      this.expressEngine(filePath, options, (err, html) => {
-        // Share the result of the render with all awaiting requests for the same key:
-
-        // Note: we access the Map at the moment of the render finished (don't store value in a local variable),
-        //       because in the meantime something might have deleted the value (i.e. when `maxRenderTime` passed).
-        this.renderCallbacks.get(renderingKey)?.forEach((cb) => cb(err, html)); // pass the shared result to all waiting rendering callbacks
-        this.renderCallbacks.delete(renderingKey);
+    if (!this.renderingCache.isRendering(renderingKey)) {
+      this.startRender({
+        filePath,
+        options,
+        request,
+        renderCallback: (err, html) => {
+          // Share the result of the render with all awaiting requests for the same key:
+
+          // Note: we access the Map at the moment of the render finished (don't store value in a local variable),
+          //       because in the meantime something might have deleted the value (i.e. when `maxRenderTime` passed).
+          this.renderCallbacks
+            .get(renderingKey)
+            ?.forEach((cb) => cb(err, html)); // pass the shared result to all waiting rendering callbacks
+          this.renderCallbacks.delete(renderingKey);
+        },
       });
     }
 
     const renderCallbacksCount = this.renderCallbacks.get(renderingKey)?.length;
     this.log(
-      `${renderCallbacksCount} requests waiting for the render to complete (${request?.originalUrl})`
+      `${renderCallbacksCount} rendering callbacks are waiting for the render to complete (${request?.originalUrl})`
```

Comment:
I would avoid mentioning "rendering callbacks", as it's an implementation detail.
From customer's perspective, _requests_ are waiting for the render.

Reply:
I agree "rendering callbacks" is an implementation detail. And it's not ideal...
But saying "requests" would  not be correct. The Express.js Request is no longer waiting, since we respond to it. Some requests might have been responded to (because of their individual timeout and their fallback to CSR).

Ideally, we would decrement the logged counter whenever a Request times out, but we keep all the renders callbacks until the render completes.

Reply:
Works, let's go with it.